vowpalwabbit/io_buf.h

```diff
@@ -10,54 +10,105 @@
 #include <vector>
 #include <memory>
 #include <cassert>
+#include <cstdlib>
 
 #include "v_array.h"
 #include "hash.h"
-#include "vw_exception.h"
 #include "io/io_adapter.h"
 
+#ifndef VW_NOEXCEPT
+#  include "vw_exception.h"
+#endif
+
 /* The i/o buffer can be conceptualized as an array below:
-**  _______________________________________________________________________________________
-** |__________|__________|__________|__________|__________|__________|__________|__________|   **
-** space.begin           space.head             space.end                       space.endarray **
+**  _________________________________________________________________
+** |__________|__________|__________|__________|__________|__________|
+** _buffer._begin        head       _buffer._end                     _buffer._end_array
 **
-** space.begin     = the beginning of the loaded values in the buffer
-** space.head      = the end of the last-read point in the buffer
-** space.end       = the end of the loaded values from file
-** space.endarray  = the end of the allocated space for the array
+** _buffer._begin     = the beginning of the loaded values in the buffer
+** head               = the end of the last-read point in the buffer
+** _buffer._end       = the end of the loaded values from file
+** _buffer._end_array = the end of the allocated _buffer for the array
 **
 ** The values are ordered so that:
-** space.begin <= space.head <= space.end <= space.endarray
+** _buffer._begin <= head <= _buffer._end <= _buffer._end_array
 **
-** Initially space.begin == space.head since no values have been read.
+** Initially _buffer._begin == head since no values have been read.
 **
-** The interval [space.head, space.end] may be shifted down to space.begin
+** The interval [head, _buffer._end] may be shifted down to _buffer._begin
 ** if the requested number of bytes to be read is larger than the interval size.
 ** This is done to avoid reallocating arrays as much as possible.
 */
 
 class io_buf
 {
+  struct internal_buffer
+  {
+    char* _begin = nullptr;
+    char* _end = nullptr;
+    char* _end_array = nullptr;
+
+    ~internal_buffer() { std::free(_begin); }
+
+    void realloc(size_t new_capacity)
+    {
+      // This specific internal buffer should only ever grow.
+      assert(new_capacity >= capacity());
+      const auto old_size = size();
+      char* temp = reinterpret_cast<char*>(std::realloc(_begin, sizeof(char) * new_capacity));
+      if (temp == nullptr)
+      {
+        // _begin still needs to be freed but the destructor will do it.
+        THROW_OR_RETURN("realloc of " << new_capacity << " failed in resize().  out of memory?");
+      }
+      else
+      {
+        _begin = temp;
+      }
+      _end = _begin + old_size;
+      _end_array = _begin + new_capacity;
+      memset(_end, 0, sizeof(char) * (_end_array - _end));
+    }
+
+    void shift_to_front(char* new_begin)
```

Comment:
`new_begin` is a confusing name. I'd say either rename it to something like `head_ptr` or maybe move the pointer inside the buffer object?